File: src/types/pack.ts
export type PackKind = 'pack' | 'project';

export interface ProjectFolderMeta {
  title: string;
  keyNote: string | null;
  keyScale: string | null;
  altKeyNote: string | null;
  altKeyScale: string | null;
  bpm: number | null;
  /** ISO 8601 date string, YYYY-MM-DD. */
  deadline: string | null;
}

export interface Pack {
  id: string;
  name: string;
  rootPath: string;
  vendor: string | null;
  genre: string | null;
  color: string | null;
  assetCount: number;
  addedAt: number;
  updatedAt: number;
  kind: PackKind;
  projectMeta: ProjectFolderMeta | null;
}

export interface WatchedFolder {
  id: string;
  path: string;
  isActive: boolean;
  addedAt: number;
  kind: PackKind;
}

export interface ScanStats {
  totalFiles: number;
  skipped: number;
  durationMs: number;
}

export interface ScanProgress {
  total: number;
  indexed: number;
  queued: number;
  isScanning: boolean;
}

export interface ReconcileReport {
  newFiles: number;
  missingFiles: number;
  durationMs: number;
  packsRemoved: number;
}

export interface CleanCacheReport {
  missingDeleted: number;
  packsDeleted: number;
}

export interface HardCleanReport {
  assetsDeleted: number;
  packsDeleted: number;
}

export interface FolderInfo {
  path: string;
  totalSizeBytes: number;
  fileCount: number;
  assetCount: number;
}

export interface CategoryStat {
  count: number;
  sizeBytes: number;
}

export interface SubfolderSummary {
  name: string;
  path: string;
  fileCount: number;
  sizeBytes: number;
}

export interface ProjectInfo {
  path: string;
  totalSizeBytes: number;
  fileCount: number;
  audio: CategoryStat;
  midi: CategoryStat;
  preset: CategoryStat;
  project: CategoryStat;
  video: CategoryStat;
  image: CategoryStat;
  other: CategoryStat;
  backupCount: number;
  backupSizeBytes: number;
  subfolders: SubfolderSummary[];
}

export interface TreeNode {
  name: string;
  path: string;
  assetCount: number;
  children: TreeNode[];
}

File: src-tauri/src/models/pack.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Pack {
    pub id: String,
    pub name: String,
    pub root_path: String,
    pub vendor: Option<String>,
    pub genre: Option<String>,
    pub color: Option<String>,
    pub asset_count: i64,
    pub added_at: i64,
    pub updated_at: i64,
    /// "pack" (default sample pack) or "project" (DAW project folder).
    pub kind: String,
    /// Parsed project folder metadata (title/key/bpm/deadline) — only set when kind = "project".
    pub project_meta: Option<crate::metadata::project_folder_parser::ProjectFolderMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WatchedFolder {
    pub id: String,
    pub path: String,
    pub is_active: bool,
    pub added_at: i64,
    pub kind: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ScanStats {
    pub total_files: usize,
    pub skipped: usize,
    pub duration_ms: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ScanProgress {
    pub total: usize,
    pub indexed: usize,
    pub queued: usize,
    pub is_scanning: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ReconcileReport {
    pub new_files: usize,
    pub missing_files: usize,
    pub duration_ms: u64,
    pub packs_removed: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TreeNode {
    pub name: String,
    pub path: String,
    pub asset_count: u32,
    pub children: Vec<TreeNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CleanCacheReport {
    pub missing_deleted: usize,
    pub packs_deleted: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct HardCleanReport {
    pub assets_deleted: usize,
    pub packs_deleted: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderInfo {
    pub path: String,
    pub total_size_bytes: u64,
    pub file_count: u64,
    pub asset_count: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CategoryStat {
    pub count: u64,
    pub size_bytes: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubfolderSummary {
    pub name: String,
    pub path: String,
    pub file_count: u64,
    pub size_bytes: u64,
}

/// Aggregated information about a project folder for the Info sidebar.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectInfo {
    pub path: String,
    pub total_size_bytes: u64,
    pub file_count: u64,
    pub audio: CategoryStat,
    pub midi: CategoryStat,
    pub preset: CategoryStat,
    pub project: CategoryStat,
    pub video: CategoryStat,
    pub image: CategoryStat,
    pub other: CategoryStat,
    pub backup_count: u64,
    pub backup_size_bytes: u64,
    pub subfolders: Vec<SubfolderSummary>,
}
